File: buffer/cpu/src/jvmMain/java/com/wsr/cpu/JMatMul.java
package com.wsr.cpu;

import java.nio.ByteBuffer;

class JMatMul {
    public native void inner(
            ByteBuffer x,
            ByteBuffer y,
            int b,
            ByteBuffer result
    );

    public native void matMulD1ToD2(
            ByteBuffer x,
            ByteBuffer y,
            boolean transY,
            int n,
            int k,
            ByteBuffer result
    );

    public native void matMulD2ToD1(
            ByteBuffer x,
            boolean transX,
            ByteBuffer y,
            int m,
            int k,
            ByteBuffer result
    );

    public native void matMulD2ToD2(
            ByteBuffer x,
            boolean transX,
            ByteBuffer y,
            boolean transY,
            int m,
            int n,
            int k,
            int b,
            ByteBuffer result
    );
}

File: buffer/cpu/src/jvmAndroidMain/kotlin/com/wsr/knist/cpu/linalg/JMatMul.kt
package com.wsr.knist.cpu.linalg

object JMatMul {
    external fun inner(x: Long, y: Long, b: Int, result: Long)

    external fun matMulD1ToD2(x: Long, y: Long, transY: Boolean, n: Int, k: Int, result: Long)

    external fun matMulD2ToD1(x: Long, transX: Boolean, y: Long, m: Int, k: Int, result: Long)

    external fun matMulD2ToD2(
        x: Long,
        transX: Boolean,
        y: Long,
        transY: Boolean,
        m: Int,
        n: Int,
        k: Int,
        b: Int,
        result: Long,
    )
}
